lib/sudo-pam/src/rpassword.rs

```diff
@@ -0,0 +1,105 @@
+/// - rtool_box::print_tty was inlined
+/// - SafeString was removed and replaced with more general Secure<> type (and moved to cutils);
+///   also, the original code actually allowed the string to quickly escape the security net
+/// - instead of String, password are read as [u8]
+/// - this also removes the need for 'fix_line_issues', since we know we read a \n
+use std::io::{self, BufRead, Write};
+use std::mem;
+use std::os::unix::io::AsRawFd;
+
+use libc::{c_int, tcsetattr, termios, ECHO, ECHONL, TCSANOW};
+
+use sudo_cutils::Secure;
+
+struct HiddenInput {
+    fd: i32,
+    term_orig: termios,
+}
+
+impl HiddenInput {
+    fn new(fd: i32) -> io::Result<HiddenInput> {
+        // Make two copies of the terminal settings. The first one will be modified
+        // and the second one will act as a backup for when we want to set the
+        // terminal back to its original state.
+        let mut term = safe_tcgetattr(fd)?;
+        let term_orig = safe_tcgetattr(fd)?;
+
+        // Hide the password. This is what makes this function useful.
+        term.c_lflag &= !ECHO;
+
+        // But don't hide the NL character when the user hits ENTER.
+        term.c_lflag |= ECHONL;
+
+        // Save the settings for now.
+        io_result(unsafe { tcsetattr(fd, TCSANOW, &term) })?;
+
+        Ok(HiddenInput { fd, term_orig })
+    }
+}
+
+impl Drop for HiddenInput {
+    fn drop(&mut self) {
+        // Set the the mode back to normal
+        unsafe {
+            tcsetattr(self.fd, TCSANOW, &self.term_orig);
+        }
+    }
+}
+
+/// Turns a C function return into an IO Result
+fn io_result(ret: c_int) -> std::io::Result<()> {
+    match ret {
+        0 => Ok(()),
+        _ => Err(std::io::Error::last_os_error()),
+    }
+}
+
+fn safe_tcgetattr(fd: c_int) -> std::io::Result<termios> {
+    let mut term = mem::MaybeUninit::<termios>::uninit();
+    io_result(unsafe { ::libc::tcgetattr(fd, term.as_mut_ptr()) })?;
+    Ok(unsafe { term.assume_init() })
+}
+
+/// Reads a password from the TTY
+pub fn read_password() -> std::io::Result<Secure<Vec<u8>>> {
+    let tty = std::fs::File::open("/dev/tty")?;
+    let fd = tty.as_raw_fd();
+    let mut reader = io::BufReader::new(tty);
+
+    read_password_from_fd_with_hidden_input(&mut reader, fd)
+}
+
+/// Reads a password from a given file descriptor
+fn read_password_from_fd_with_hidden_input(
+    reader: &mut impl BufRead,
+    fd: i32,
+) -> std::io::Result<Secure<Vec<u8>>> {
+    let mut password = Vec::new();
+
+    let hidden_input = HiddenInput::new(fd)?;
+
+    let status = reader.read_until(0x0a, &mut password);
+    let password = Secure::new(password);
+
+    std::mem::drop(hidden_input);
+
+    status.map(|_| password)
+}
+
+/// Prompts on the TTY and then reads a password from TTY
+pub fn prompt_password(prompt: impl ToString) -> std::io::Result<Secure<Vec<u8>>> {
+    let mut stream = std::fs::OpenOptions::new().write(true).open("/dev/tty")?;
```

Comment:
Feels like we can write a better version here that only needs to open the TTY once.

Reply:
Yes.